# U.S. Medical Insurance Costs

## 1. Data Loading & Exploratory Data Analysis

In this section, we load `insurance.csv` into a Pandas DataFrame and inspect:
* General structure and non-null values
* Core numerical distributions (`age`, `bmi`, `children`, `charges`)
* Categorical distributions (`sex`, `smoker`, `region`)

In [28]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set visual theme
sns.set_theme(style="whitegrid")

# Load the dataset
df = pd.read_csv('insurance.csv')

# Display shape and column summary
print(f"Dataset Dimensions: {df.shape[0]} rows, {df.shape[1]} columns")
print("\n--- Data Overview ---")
df.info()

print("\n--- Missing Values ---")
print(df.isnull().sum())

print("\n--- First 5 Rows ---")
display(df.head())

print("\n--- Numerical Summary ---")
display(df.describe().round(2))

Dataset Dimensions: 1338 rows, 7 columns

--- Data Overview ---
<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 73.3 KB

--- Missing Values ---
age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

--- First 5 Rows ---


,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520



--- Numerical Summary ---


,age,bmi,children,charges
count,1338.00,1338.00,1338.00,1338.00
mean,39.21,30.66,1.09,13270.42
std,14.05,6.10,1.21,12110.01
min,18.00,15.96,0.00,1121.87
25%,27.00,26.30,0.00,4740.29
50%,39.00,30.40,1.00,9382.03
75%,51.00,34.69,2.00,16639.91
max,64.00,53.13,5.00,63770.43


# Project Scope: U.S. Medical Insurance Costs Analysis

## 1. Project Goals
The main goal of this project is to analyze individual medical insurance costs (`charges`) to identify key demographic and health risk factors that drive healthcare expenditure.

## 2. Key Questions & Hypotheses
1. **Smoker Status:** How significantly do healthcare costs differ between smokers and non-smokers?
   * *Hypothesis:* Smokers will incur substantially higher charges than non-smokers due to health risk factors.
2. **BMI Interaction:** Does Body Mass Index (BMI) amplify medical costs independently, or is its effect compounded when paired with smoking?
   * *Hypothesis:* Individuals with high BMI ($\ge 30$) who smoke will experience the highest average charges in the dataset.
3. **Age & Cost Correlation:** How steadily do insurance costs scale with age?
   * *Hypothesis:* Charges will show a positive linear trend with age across all demographics.
4. **Regional Differences:** Does geographic region (`northeast`, `northwest`, `southeast`, `southwest`) significantly impact medical insurance charges?
   * *Hypothesis:* Regional costs will remain mostly balanced, though localized factors may create minor variances.
5. **Family Size / Children:** Does the number of children/dependents covered correlate with increased individual charges?

## 3. Dataset Characteristics
* **Source:** `insurance.csv`
* **Size:** 1,338 records, 7 attributes.
* **Integrity:** Clean dataset with 0 missing or null values.
* **Feature Types:**
  * **Numerical:** `age` (discrete), `bmi` (continuous), `children` (discrete), `charges` (continuous)
  * **Categorical:** `sex`, `smoker`, `region`

## 4. Planned Analytical Workflow
1. **Exploratory Data Analysis (EDA):** Generate descriptive statistics (mean, median, standard deviation) for core parameters.
2. **Segment Analysis & Grouping:** Group dataset by categorical variables (`smoker`, `region`, `sex`) to evaluate aggregated charge differences.
3. **Correlation & Feature Engineering:** Analyze correlations between numerical features (`age`, `bmi`, `children`) and `charges`.
4. **Data Visualization:** Construct box plots, scatter plots, and heatmaps using `seaborn` and `matplotlib` to validate key insights visually.

In [29]:
# Compute summary statistics mapped in the project scope
smoker_analysis = df.groupby('smoker')['charges'].agg(['count', 'mean', 'median', 'std']).round(2)
region_analysis = df.groupby('region')['charges'].agg(['count', 'mean', 'median', 'std']).round(2)

print("--- Smoker vs Non-Smoker Charges ---")
print(smoker_analysis)

print("\n--- Regional Breakdown ---")
print(region_analysis)

--- Smoker vs Non-Smoker Charges ---
        count      mean    median       std
smoker                                     
no       1064   8434.27   7345.41   5993.78
yes       274  32050.23  34456.35  11541.55

--- Regional Breakdown ---
           count      mean    median       std
region                                        
northeast    324  13406.38  10057.65  11255.80
northwest    325  12417.58   8965.80  11072.28
southeast    364  14735.41   9294.13  13971.10
southwest    325  12346.94   8798.59  11557.18


In [30]:
import pandas as pd

# Load the dataset
df = pd.read_csv('insurance.csv')

# Verify dimensions, data types, and check for missing values
print(f"Dataset Shape: {df.shape[0]} rows, {df.shape[1]} columns\n")
print("--- Data Information ---")
df.info()

print("\n--- Missing Values ---")
print(df.isnull().sum())

print("\n--- First 5 Rows ---")
display(df.head())

print("\n--- Numerical Summary Statistics ---")
display(df.describe().round(2))

Dataset Shape: 1338 rows, 7 columns

--- Data Information ---
<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 73.3 KB

--- Missing Values ---
age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

--- First 5 Rows ---


,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520



--- Numerical Summary Statistics ---


,age,bmi,children,charges
count,1338.00,1338.00,1338.00,1338.00
mean,39.21,30.66,1.09,13270.42
std,14.05,6.10,1.21,12110.01
min,18.00,15.96,0.00,1121.87
25%,27.00,26.30,0.00,4740.29
50%,39.00,30.40,1.00,9382.03
75%,51.00,34.69,2.00,16639.91
max,64.00,53.13,5.00,63770.43


In [31]:
# Save each dataset feature into its own dedicated variable
ages = df['age']
sexes = df['sex']
bmis = df['bmi']
num_children = df['children']
smoker_statuses = df['smoker']
regions = df['region']
charges = df['charges']

# Confirm variables are populated by checking lengths and sample values
print(f"Total entries per variable: {len(ages)}")
print(f"Sample - First Age: {ages[0]}, First Charge: ${charges[0]:.2f}")

Total entries per variable: 1338
Sample - First Age: 19, First Charge: $16884.92


In [32]:
class InsuranceAnalysis:
    def __init__(self, ages, sexes, bmis, num_children, smoker_statuses, regions, charges):
        self.ages = ages
        self.sexes = sexes
        self.bmis = bmis
        self.num_children = num_children
        self.smoker_statuses = smoker_statuses
        self.regions = regions
        self.charges = charges

    def analyze_ages(self):
        """Calculates average, minimum, and maximum age."""
        avg_age = sum(self.ages) / len(self.ages)
        return {"mean_age": round(avg_age, 2), "min_age": min(self.ages), "max_age": max(self.ages)}

    def analyze_regions(self):
        """Calculates counts and percentages for each unique region."""
        total = len(self.regions)
        counts = {}
        for region in self.regions:
            counts[region] = counts.get(region, 0) + 1
        return {region: {"count": count, "percentage": round((count / total) * 100, 2)} 
                for region, count in counts.items()}

    def analyze_smoker_charges(self):
        """Compares average charges between smokers and non-smokers."""
        smoker_costs = [c for c, s in zip(self.charges, self.smoker_statuses) if s == 'yes']
        non_smoker_costs = [c for c, s in zip(self.charges, self.smoker_statuses) if s == 'no']

        avg_smoker = sum(smoker_costs) / len(smoker_costs)
        avg_non_smoker = sum(non_smoker_costs) / len(non_smoker_costs)

        return {
            "avg_smoker_cost": round(avg_smoker, 2),
            "avg_non_smoker_cost": round(avg_non_smoker, 2),
            "difference": round(avg_smoker - avg_non_smoker, 2)
        }

    def create_patient_dictionary(self):
        """Creates a consolidated dictionary for all patient records."""
        patient_data = {}
        for i in range(len(self.ages)):
            patient_data[i] = {
                "age": self.ages[i],
                "sex": self.sexes[i],
                "bmi": self.bmis[i],
                "children": self.num_children[i],
                "smoker": self.smoker_statuses[i],
                "region": self.regions[i],
                "charges": self.charges[i]
            }
        return patient_data

    def generate_summary_dictionary(self):
        """Compiles all high-level metrics into a nested dictionary."""
        return {
            "demographics": {
                "total_patients": len(self.ages),
                "average_age": self.analyze_ages()["mean_age"],
                "regional_breakdown": self.analyze_regions()
            },
            "financials": {
                "smoker_impact": self.analyze_smoker_charges(),
                "average_overall_charge": round(sum(self.charges) / len(self.charges), 2)
            }
        }

    def analyze_feature_influence(self):
        """Calculates Pearson correlation coefficient for numerical/binary features against charges."""
        smoker_num = [1 if s == 'yes' else 0 for s in self.smoker_statuses]
        sex_num = [1 if s == 'male' else 0 for s in self.sexes]
        
        def pearson_r(x, y):
            n = len(x)
            mean_x, mean_y = sum(x)/n, sum(y)/n
            cov = sum((xi - mean_x) * (yi - mean_y) for xi, yi in zip(x, y))
            std_x = (sum((xi - mean_x)**2 for xi in x)) ** 0.5
            std_y = (sum((yi - mean_y)**2 for yi in y)) ** 0.5
            return round(cov / (std_x * std_y), 4)

        return {
            "smoker_correlation": pearson_r(smoker_num, self.charges),
            "age_correlation": pearson_r(self.ages, self.charges),
            "bmi_correlation": pearson_r(self.bmis, self.charges),
            "children_correlation": pearson_r(self.num_children, self.charges),
            "sex_correlation": pearson_r(sex_num, self.charges)
        }

# Execute class methods
analysis = InsuranceAnalysis(ages, sexes, bmis, num_children, smoker_statuses, regions, charges)

print("Age Statistics:", analysis.analyze_ages())
print("\nRegional Distribution:", analysis.analyze_regions())
print("\nSmoker vs Non-Smoker Costs:", analysis.analyze_smoker_charges())


Age Statistics: {'mean_age': 39.21, 'min_age': 18, 'max_age': 64}

Regional Distribution: {'southwest': {'count': 325, 'percentage': 24.29}, 'southeast': {'count': 364, 'percentage': 27.2}, 'northwest': {'count': 325, 'percentage': 24.29}, 'northeast': {'count': 324, 'percentage': 24.22}}

Smoker vs Non-Smoker Costs: {'avg_smoker_cost': 32050.23, 'avg_non_smoker_cost': 8434.27, 'difference': 23615.96}


## Key Findings & Project Conclusions

1. **Average Age**: The mean age of policyholders in this dataset is **39.2 years** (ranging from 18 to 64).
2. **Smoker Cost Disparity**: 
   * **Smoker Average Cost**: ~$32,050.23
   * **Non-Smoker Average Cost**: ~$8,434.27
   * **Key Insight**: Smoking is the single largest contributing factor to elevated medical costs, increasing average charges by **over 280%** (~$23,615 difference).
3. **Regional Distribution**:
   * The dataset is evenly split across four primary regions (~25% each), with the **Southeast** having a slightly higher representation (~27%).

In [33]:
def generate_summary_dictionary(self):
    """Compiles all high-level metrics into a nested dictionary."""
    return {
        "demographics": {
            "total_patients": len(self.ages),
            "average_age": self.analyze_ages()["mean_age"],
            "regional_breakdown": self.analyze_regions()
        },
        "financials": {
            "smoker_impact": self.analyze_smoker_charges(),
            "average_overall_charge": round(sum(self.charges) / len(self.charges), 2)
        }
    }

# Add method call
summary_dict = analysis.generate_summary_dictionary()
import pprint
pprint.pprint(summary_dict)

{'demographics': {'average_age': 39.21,
                  'regional_breakdown': {'northeast': {'count': 324,
                                                       'percentage': 24.22},
                                         'northwest': {'count': 325,
                                                       'percentage': 24.29},
                                         'southeast': {'count': 364,
                                                       'percentage': 27.2},
                                         'southwest': {'count': 325,
                                                       'percentage': 24.29}},
                  'total_patients': 1338},
 'financials': {'average_overall_charge': 13270.42,
                'smoker_impact': {'avg_non_smoker_cost': 8434.27,
                                  'avg_smoker_cost': 32050.23,
                                  'difference': 23615.96}}}


In [34]:
# Call the feature influence method
feature_correlations = analysis.analyze_feature_influence()

import pprint
print("Correlation with Medical Charges (Pearson r):")
pprint.pprint(feature_correlations)


Correlation with Medical Charges (Pearson r):
{'age_correlation': 0.299,
 'bmi_correlation': 0.1983,
 'children_correlation': 0.068,
 'sex_correlation': 0.0573,
 'smoker_correlation': 0.7873}


## Extension 2: Predictive Insights & Feature Importance

Based on our correlation calculations, we can rank the features by their influence on individual medical charges:

1. **Smoker Status ($r \approx 0.7873$) — Most Influential**:
   * Smoking status shows a very strong positive correlation with medical charges. 
   * It is the single biggest predictor of higher insurance costs.

2. **Age ($r \approx 0.2990$) — Moderate Influence**:
   * Age has a steady positive relationship with charges. As policyholders age, medical costs gradually increase regardless of other factors.

3. **BMI ($r \approx 0.1983$) — Conditional Influence**:
   * While overall linear correlation with charges is mild, BMI exhibits a dramatic compounding effect when combined with smoking status (e.g., smokers with a high BMI experience exponentially higher charges).

4. **Children ($r \approx 0.0680$) & Sex ($r \approx 0.0573$) — Negligible Direct Impact**:
   * Neither the number of dependents nor sex shows a strong direct linear correlation with insurance costs.

In [35]:
import pandas as pd

# Load dataset to inspect sampling distributions
df = pd.read_csv('insurance.csv')

# Categorize BMI according to standard WHO ranges
df['bmi_category'] = pd.cut(
    df['bmi'], 
    bins=[0, 18.5, 24.9, 29.9, 100], 
    labels=['Underweight', 'Normal', 'Overweight', 'Obese']
)

# Cross-tabulate Smoker Status vs BMI Category
bmi_smoker_cross = pd.crosstab(df['bmi_category'], df['smoker'], margins=True)
print("Subgroup Counts (BMI Category vs Smoker):")
display(bmi_smoker_cross)

# Region representation
print("\nRegional Distribution:")
print(df['region'].value_counts(normalize=True).round(4) * 100)

Subgroup Counts (BMI Category vs Smoker):


smoker,no,yes,All
bmi_category,,,
Underweight,16,5,21
Normal,171,50,221
Overweight,308,72,380
Obese,569,147,716
All,1064,274,1338



Regional Distribution:
region
southeast    27.20
southwest    24.29
northwest    24.29
northeast    24.22
Name: proportion, dtype: float64


## Extension 3: Exploring Dataset Bias & Limitations

When evaluating this dataset for real-world application or predictive risk modeling, several structural biases and sampling limitations must be considered:

### 1. Severe Subgroup Underrepresentation (Sampling Bias)
* **Underweight Smoker Representation:** Out of 1,338 records, only **21 individuals** are classified as "Underweight" ($\text{BMI} < 18.5$), and only **5** of those are smokers. 
* **Impact:** Statistical estimates or predictive models trained on this data will have low statistical power and high variance when predicting outcomes for underweight individuals.

### 2. Omission of Critical Health & Clinical Metrics (Confounding Variables)
* **Missing Clinical Markers:** The dataset contains no medical history, blood pressure readings, cholesterol levels, family history of chronic illnesses, pre-existing conditions, or lifestyle factors (e.g., alcohol consumption, exercise routines).
* **Impact:** Attributing higher insurance costs solely to age, BMI, and smoking status risks oversimplifying risk models. Two individuals with identical age and BMI could have radically different medical expenses depending on underlying conditions not captured in this file.

### 3. Coarse Geographic & Socioeconomic Granularity
* **Broad Regional Buckets:** Location is recorded across only four broad quadrants (`northeast`, `northwest`, `southeast`, `southwest`).
* **Impact:** Healthcare pricing varies drastically by state regulations, local cost of living, urban vs. rural healthcare access, and regional hospital network monopolies. Lumping entire regions into single categories ignores micro-economic variations in medical pricing.

### 4. Selection & Survivorship Bias
* **Active Policyholders Only:** This dataset only captures individuals who currently hold medical insurance policies.
* **Impact:** Uninsured individuals—who may face higher health risks or abstain from insurance due to cost barriers—are completely excluded. Models trained on this dataset cannot be generalized to the broader national population.